### PR TITLE
doc: make developer guide great again (by making it a book)

### PR DESCRIPTION
**Pull request by xxchan**
Signed-off-by: xxchan <xxchan22f@gmail.com>I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Reorganize `developer-guide.md` into multiple chapters, and build it with `mdbook`. See the demo here: https://xxchan.me/risingwave/

The PR doesn't add new contents yet, but mainly just split it into chapters and introduce the `mdbook`.

Reasons:
- My original motivation is to add docs to explain how to test connectors (the ideas (risedev & slt) from https://github.com/risingwavelabs/risingwave/issues/12451#issuecomment-2051861048). 
  But I found `developer-guide.md` is poorly maintained for a while, and I raised https://github.com/risingwavelabs/risingwave/issues/17436. So I wanted to revive it and then create a better for the connector test doc.
- `developer-guide.md` it self is too long (read current version [here](https://github.com/risingwavelabs/risingwave/blob/3795168b11f216cb035dbd3130992d417ca6f9ea/docs/developer-guide.md)), and contains too many topics. It doesn't have good center points. It's not helpful for new developers (who just need quick start, but not advanced topics) and also experienced developers (who need more detailed doc about some complex topics, or best practices (e.g., the connector test topic. Actually we don't have such docs now)).
    - Note: So we have to split it into multiple files. `mdbook` is actually optional.
    - In the meantime, we also have other developer docs scattered around lots of places, e.g., RiseDev's doc (I believe there are many useful RiseDev features unknown to many developers).
    - After putting them in one place and with separated chapters, I believe the dev guide will be **more useful to both new and experienced developers and also more easy to maintain**. I hope it can help use **document more tribal knowledge**.
- Some other benefits of `mdbook`
    - Introducing `mdbook` is actually no much complexity. No specialized syntax, and can still be read as plain md files. Just add one `book.toml`.
    - Auto generated TOC. Auto link check.
    - Nice sidebar. 
    - More readable..?

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: docs/dev/src/SUMMARY.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+[Introduction](./intro.md)
+[Contribution Procedure](./contribution.md)
+
+---
+
+# Building and debugging RisingWave
+
+- [How to build and run RisingWave](./build-and-run/intro.md)
+- [Testing](./tests/intro.md)
```
Comment (xxchan) [inline docs/dev/src/SUMMARY.md:11]: Each of them is worth expanding into several sub-chapters later...

File: docs/dev/src/SUMMARY.md [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+- [CI](./ci.md)
+
+<!--
+
+TODO:
+
+- [RiseDev](./risedev.md)
+- [Error Handling](./error-handling.md)
+- [Develop Connector]()
```
Comment (xxchan) [inline docs/dev/src/SUMMARY.md:35]: This is the chapter I pretty want to write :)

File: docs/dev/book.toml [hunk 3]
```diff
@@ -0,0 +1,15 @@
+[book]
+authors = ["xxchan"]
```
Comment (BugenZhao) [inline docs/dev/book.toml:2]: 😄

File: docs/dev/README.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
+# RisingWave Developer guide
+
+This is the source of the RisingWave Developer guide, published at
+<https://risingwavelabs.github.io/risingwave/> (via GitHub Actions to GitHub Pages).
```
Comment (BugenZhao) [inline docs/dev/README.md:4]: May worth note that this is previously for rustdoc of all crates in RisingWave.

**Review by BugenZhao — COMMENTED** [review]
> After putting them in one place and with separated chapters, I believe the dev guide will be more useful to both new and experienced developers and also more easy to maintain.

Strongly agree on this. Experienced developers can easily focus on the new advanced articles without the hassle of checking the updates in existing docs.

File: docs/README.md [hunk 5]
```diff
@@ -4,7 +4,7 @@ This directory contains RisingWave design documents that are intended to be used
 
 ## Developer guide
 
-After you learn about the basics of RisingWave, take a look at our [developer guide](developer-guide.md). It'll help you up to speed with the development process.
+After you learn about the basics of RisingWave, take a look at our [developer guide](https://risingwavelabs.github.io/risingwave/). It'll help you up to speed with the development process.
```
Comment (kwannoel) [inline docs/README.md:7]:
```suggestion
After you learn about the basics of RisingWave, take a look at our [developer guide](https://risingwavelabs.github.io/risingwave/) to get up to speed with the development process.
```

File: docs/dev/src/ci.md [hunk 6]
```diff
@@ -0,0 +1,21 @@
+# CI
```
Comment (kwannoel) [inline docs/dev/src/ci.md:1]:
```suggestion
# Continuous Integration
```

Design-wise, it looks weird just to have `CI`.

Screenshot 2024-07-02 at 12 49 56 PM

File: docs/dev/src/SUMMARY.md [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+- [CPU Profiling Guide](./benchmark-and-profile/cpu-profiling.md)
+- [Memory (Heap) Profiling Guide](./benchmark-and-profile/memory-profiling.md)
+- [Microbench Guide](./benchmark-and-profile/microbenchmarks.md)
+
+---
+
+# Specialized topics
+
+- [CI](./ci.md)
```
Comment (kwannoel) [inline docs/dev/src/SUMMARY.md:27]:
```suggestion
- [Continuous Integration](./ci.md)
```

**Review by kwannoel — APPROVED** [review]
Rest LGTM, it will make contribution much more pleasant for both new and experienced devs. Thanks!

File: docs/dev/src/contribution.md [hunk 8]
```diff
@@ -0,0 +1,64 @@
+# Contribution guidelines
```
Comment (stdrc) [inline docs/dev/src/contribution.md:1]:
```suggestion
# Contributing Guidelines
```

File: docs/dev/book.toml [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+title = "RisingWave Dev Guide"
```
Comment (stdrc) [inline docs/dev/book.toml:3]:
```suggestion
title = "RisingWave Developer Guide"
```

File: docs/dev/src/SUMMARY.md [hunk 10]
```diff
@@ -0,0 +1,41 @@
+# Summary
+
+[Introduction](./intro.md)
+[Contribution Procedure](./contribution.md)
```
Comment (stdrc) [inline docs/dev/src/SUMMARY.md:4]:
```suggestion
[Contributing Guidelines](./contribution.md)
```

File: docs/dev/src/contribution.md
Comment (stdrc) [inline docs/dev/src/contribution.md:1]: Rename to `contributing.md`? https://docs.github.com/en/communities/setting-up-your-project-for-healthy-contributions/setting-guidelines-for-repository-contributors

File: docs/dev/src/SUMMARY.md [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+---
+
+# Building and debugging RisingWave
+
+- [How to build and run RisingWave](./build-and-run/intro.md)
```
Comment (stdrc) [inline docs/dev/src/SUMMARY.md:10]: The title looks too long to me. What about "Getting Started"/"Building and Running"?

File: docs/dev/src/SUMMARY.md [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+- [Debugging](./debugging.md)
+- [Observability](./observability.md)
+
+---
+
+# Benchmarking and Profiling
+
+- [CPU Profiling Guide](./benchmark-and-profile/cpu-profiling.md)
+- [Memory (Heap) Profiling Guide](./benchmark-and-profile/memory-profiling.md)
+- [Microbench Guide](./benchmark-and-profile/microbenchmarks.md)
```
Comment (stdrc) [inline docs/dev/src/SUMMARY.md:21]: Why append `Guide`? What about:

- CPU Profiling
- Memory (Heap) Profiling
- Microbenchmarking

File: docs/dev/src/build-and-run/intro.md
Comment (stdrc) [inline docs/dev/src/build-and-run/intro.md:1]: Why `intro.md`? What about `index.md`?

**Comment by fuyufjh** [comment]
Great work!

How about our design docs and our RFCs? Is it possible to migrate them here as well?

**Comment by xxchan** [comment]
> How about our design docs and our RFCs? Is it possible to migrate them here as well?

For design docs, I think it's good to include them. 

For RFC docs, I'm not sure. The quality (at least language) seems to fluctuate. And the number of RFCs is also large. If necessary, we can simply create an "RFC book" like https://rust-lang.github.io/rfcs/introduction.html

File: docs/dev/book.toml (hunk 9 above)
Comment (xxchan) [reply docs/dev/book.toml:3]:
```suggestion
title = "RisingWave Development Guide"
```

File: docs/dev/src/build-and-run/intro.md
Comment (xxchan) [reply docs/dev/src/build-and-run/intro.md:1]: findings:
- mdBook calls it `README.md`
- rustc-dev-guide calls it `intro.md`, or arbitrary name
- cargo calls it `index.md`

**Pull request merged**